### PR TITLE
Cleanup container LogEvent calls

### PR DESCRIPTION
**Pull request by duglin**
Move some calls to container.LogEvent down lower so that there's
less of a chance of them being missed. Also add a few more events
that appear to have been missed.

Added testcases for new events: commit, copy, resize, attach, rename, top

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

**Comment by runcom** [comment]
could we have some tests for these new events also (since we are testing much like all of the others) :angel: ?

**Comment by duglin** [comment]
@runcom added tests.

**Comment by duglin** [comment]
Added testcases for new events: commit, copy, resize, attach, rename, top

File: integration-cli/docker_cli_events_test.go [hunk 1]
```diff
@@ -563,3 +572,218 @@ func (s *DockerSuite) TestEventsStreaming(c *check.C) {
...
+
+func (s *DockerSuite) TestEventsAttach(c *check.C) {
+	since := daemonTime(c).Unix()
+
+	out, _ := dockerCmd(c, "run", "-di", "busybox", "/bin/cat")
+	cID := strings.TrimSpace(out)
+
+	cmd := exec.Command(dockerBinary, "attach", cID)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
```
Comment (runcom) [inline integration-cli/docker_cli_events_test.go]: no big deal here but c.Assert would be cool for simple err checks (below as well) maybe another cleanup pr for beginner

**Comment by runcom** [comment]
LGTM like these extra events

File: integration-cli/docker_cli_events_test.go (hunk 1 above)
Comment (duglin) [inline integration-cli/docker_cli_events_test.go]: Yea, I would normally print a nicer error msg along with it but I copy-n-pasted this code from someplace else.  I'll switch it to just an assert since we don't show a nicer msg

**Comment by jfrazelle** [comment]
LGTM, but needs rebasse ;)

**Comment by duglin** [comment]
rebased

File: daemon/container.go [hunk 2]
```diff
@@ -462,6 +469,7 @@ func (container *Container) Kill() error {
 	}
 
 	container.WaitStop(-1 * time.Second)
+	container.LogEvent("kill")
```
Comment (LK4D4) [inline daemon/container.go]: You already sent this event from `KillSig`.

File: daemon/container.go [hunk 3]
```diff
@@ -822,10 +843,15 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
...
-	return attach(&c.StreamConfig, c.Config.OpenStdin, c.Config.StdinOnce, c.Config.Tty, stdin, stdout, stderr)
+	err := attach(&c.StreamConfig, c.Config.OpenStdin, c.Config.StdinOnce, c.Config.Tty, stdin, stdout, stderr)
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (c *Container) AttachWithLogs(stdin io.ReadCloser, stdout, stderr io.Writer, logs, stream bool) error {
+	c.LogEvent("attach")
```
Comment (LK4D4) [inline daemon/container.go]: I think this should be in `c.Attach`.

File: daemon/container.go [hunk 4]
```diff
@@ -822,10 +843,15 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
 }
 
 func (c *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writer) chan error {
-	return attach(&c.StreamConfig, c.Config.OpenStdin, c.Config.StdinOnce, c.Config.Tty, stdin, stdout, stderr)
+	err := attach(&c.StreamConfig, c.Config.OpenStdin, c.Config.StdinOnce, c.Config.Tty, stdin, stdout, stderr)
```
Comment (LK4D4) [inline daemon/container.go]: I didn't understand this change. Btw it returns `chan`, so now it's super-confusing.

File: daemon/top.go [hunk 5]
```diff
@@ -68,5 +68,6 @@ func (daemon *Daemon) ContainerTop(name string, psArgs string) (*types.Container
 			}
 		}
 	}
+	container.LogEvent("top")
```
Comment (LK4D4) [inline daemon/top.go]: Pretty weird event as for me. I don't see any usage for it.

File: daemon/container.go (hunk 2 above)
Comment (duglin) [inline daemon/container.go]: Its a little hard to follow all of the various kill funcs, but I think you're right. I'll remove this one.

File: daemon/container.go (hunk 4 above)
Comment (duglin) [inline daemon/container.go]: reverted it but it shouldn't have changed anything.

File: daemon/container.go (hunk 3 above)
Comment (duglin) [inline daemon/container.go]: I put it here because I think all flows go through here. If we're not streaming then we won't call c.Attach().

File: daemon/top.go (hunk 5 above)
Comment (duglin) [inline daemon/top.go]: I added this one because my thinking was that anytime we do **anything** to/with a container we need to log it so that people can debug stuff.  In this case we started another process in the container so someone watching the event log might need to know that info.

**Comment by LK4D4** [comment]
There is some inconsistencies, some events spawned before action, some after. So, some will be spawned every time, some only if there is no error on execution.
@duglin Do you want to improve this or merge it as is?

**Comment by duglin** [comment]
@LK4D4 if there are inconsistencies let me try to fix them as part of this.  But, let's first agree on the over flow around how this should look.  My thoughts:
- in general the call to container.LogEvent() should be after the action is completed successfully
- the exception to that rule is when the action is a long-running action, like attach()
- in the case of an error, the LogEvent will not appear in the event stream, but instead there should be a logrus ERROR instead

Sound ok?

I have more thoughts on how to change our error logging but I need #9509 to be done first.

**Comment by duglin** [comment]
@LK4D4 while waiting for your thoughts, I checked, and I think there might be just a small number that would need to be tweaked, most seem to log the event at the end of processing - so that's good.

**Comment by duglin** [comment]
@LK4D4 I went ahead and made the changes  :-)   see what you think.

**Comment by LK4D4** [comment]
@duglin will look in ten minutes, thanks!

File: daemon/container.go [hunk 6]
```diff
@@ -836,6 +858,8 @@ func (c *Container) AttachWithLogs(stdin io.ReadCloser, stdout, stderr io.Writer
 			logrus.Errorf("Reading logs not implemented for driver %s", c.LogDriverType())
 		} else {
 			dec := json.NewDecoder(cLog)
+			c.LogEvent("attach")
```
Comment (LK4D4) [inline daemon/container.go]: This block isn't returning from function. So, I think it's enough to Log just before `if stream`. Because there showing all logs is finished and we're about to start operation with indefinite length, if `stream` is `false` then logging this event will be ok too as if we finished to stream immediately.

Comment (duglin) [inline daemon/container.go]: ok - I wasn't sure if this JSONLog copier would block or not, guess not.

**Comment by duglin** [comment]
@LK4D4 ok fixed it

**Comment by LK4D4** [comment]
LGTM
